File: code/ass1/src/main/scala/Gibbs.scala
package ass1

object Gibbs {
  abstract class State{ def update: State }

  // For large B, print out results instead of storing a list.
  def sample[T <: State](init: T, B: Int, burn: Int, printEvery: Int = 10) = {
    def loop(S: List[T], i: Int): List[T] = {
      if (i < B + burn) {
        if (i % printEvery == 0) print("\rProgress: " + i +"/"+ (B+burn))
        val newState = S.head.update.asInstanceOf[T] :: S
        loop(newState, i+1)
      } else {
        print("\rProgress: " + i + "/" + (B + burn) + "\n")
        S
      }
    }
    loop(List(init),0).take(B)
  }

}

File: cytof/src/model3/cytof3/R/gibbs.R
gibbs = function(init, update, B, burn, print_every=0) {
  ### Preallocated Memory
  out = lapply(as.list(1:B), function(b) init)

  ### Gibbs Update
  for (i in 1:(B+burn)) {
    if (i - burn <= 1) {
      update(out[[1]])
    } else {
      out[[i - burn]] = out[[i - burn - 1]]
      update(out[[i - burn]])
    }

    ### Print progress
    if (print_every > 0 && i %% print_every == 0) {
      cat("\rProgress: ", i, " / ", B+burn)
    }
  }
  cat("\n")

  out
}

